# ¡Hola Jorge!

Mi nombre es Ezequiel Ferrario, soy code reviewer en Tripleten y tengo el agrado de revisar el proyecto que entregaste.

Para simular la dinámica de un ambiente de trabajo, si veo algún error, en primer instancia solo los señalaré, dándote la oportunidad de encontrarlos y corregirlos por tu cuenta. En un trabajo real, el líder de tu equipo hará una dinámica similar. En caso de que no puedas resolver la tarea, te daré una información más precisa en la próxima revisión.

Encontrarás mis comentarios más abajo - **por favor, no los muevas, no los modifiques ni los borres**.

¿Cómo lo voy a hacer? Voy a leer detenidamente cada una de las implementaciones que has llevado a cabo para cumplir con lo solicitado. Verás los comentarios de esta forma:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta. Se aceptan uno o dos comentarios de este tipo en el borrador, pero si hay más, deberá hacer las correcciones. Es como una tarea de prueba al solicitar un trabajo: muchos pequeños errores pueden hacer que un candidato sea rechazado.
</div>

<div class="alert alert-block alert-danger">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>

Puedes responderme de esta forma:

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>

Hola, muchas gracias por tus comentarios y la revisión.
</div>

¡Empecemos!

# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)


<div class="alert alert-block alert-success">

<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido, muy bien.
</div>

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [119]:
import pandas as pd # Importar pandas


Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [120]:
df = pd.read_csv('/datasets/music_project_en.csv') # Leer el archivo y almacenarlo en df


Muestra las 10 primeras filas de la tabla:

In [93]:
print(df.head(10))# Obtener las 10 primeras filas de la tabla df


     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5



<div class="alert alert-block alert-success">

<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido.
</div>

OK LISTO

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [94]:
print(df.info())# Obtener la información general sobre nuestros datos


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Hal final de los encabezados tambien hay mayusculas y debe ser mayusculas`.




### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

1. Tematica de musica escuchada por usuarios de Shelbyville y Springfield. anlisis de escuchar la música en esas poblaciones.
2. Necistamos mas informacion para tener mas discrepancia y dar un buen veredicto.
3. Si se observa, duplicados(65079 y non-null) o tipos de datos incorrectos(non-null) y en los encabezados de "object" hay que correguilos.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Esta bien. Siempre que hagamos este tipo de observaciones, extendernos en el texto estara muy bien ya que agregara valor a nuestro trabajo.</div>

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [121]:
#import pandas as pd
#df = pd.read_csv('/datasets/music_project_en.csv')
# Muestra los nombres de las columnas
columns_new = {
    '  userID':'user_id',
    'Track':'track',
    '  City  ':'city',# tiene 2 espacio a la derechea 
    'Day':'day',
}

df.rename(columns=columns_new, inplace=True)

print(df.columns)    


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')




<div class="alert alert-block alert-success">

<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido, perfecto.
</div>

In [ ]:
LISTO GRACIAS

Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [1]:
#import pandas as pd
#df = pd.read_csv('/datasets/music_project_en.csv')

new_col_names = []
# Bucle en los encabezados poniendo todo en minúsculas
for old_name in df.columns:
    # Bucle en los encabezados poniendo todo en minúsculas
    name_lowered = old_name.lower()
    new_col_names.append(name_lowered)

df.columns = new_col_names
#df.rename(columns=columns_new, inplace=True) 

print(df.columns)



NameError: name 'df' is not defined

<div class="alert alert-block alert-danger">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

No importes pandas ya que lo has hecho al principio, no constituye una buena practica hacerla todo el tiempo.
Otra cosa que no se debe hacer es cargar nuevamente el dataset(con el read). Ya que estaras todo el tiempo cancelando tu progreso con el mismo, en el momento que lo haces reinicias todos los cambios, por esto es que se carga solo al principio.
</div>

<div class="alert alert-block alert-success">

<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido
</div>

LISTO

Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [97]:
#import pandas as pd
#df = pd.read_csv('/datasets/music_project_en.csv')

new_col_names = []
# Bucle en los encabezados eliminando los espacios
for old_name in df.columns:
   
    name_stripped = old_name.strip()
    new_col_names.append(name_stripped)
    

# Reemplaza los nombres anteriores por los nuevos
df.columns = new_col_names# escribe tu código aquí
#df.rename(columns = columns_new, inplace=True)

print(df.columns)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [98]:
#import pandas as pd
#df = pd.read_csv('/datasets/music_project_en.csv')
# Muestra los nombres de las columnas

# escribe tu código aquí
df.rename(columns={'userid': 'user_id'}, inplace=True)
#df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')
    

#df.rename(columns = columns_new, inplace=True)

print(df.columns)    


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Comprueba el resultado. Muestra los encabezados una vez más:

In [123]:
#import pandas as pd
#df = pd.read_csv('/datasets/music_project_en.csv')

new_col_names = []
for old_name in df.columns:
    # Primero, elimina los espacios al principio y al final
    name_stripped = old_name.strip()
    # Luego, pon todas las letras en minúsculas
    name_lowered = name_stripped.lower()
    # Por último, reemplaza los espacios entre palabras por guiones bajos
    name_no_spaces = name_lowered.replace(' ', '_')
    # Agrega el nuevo nombre a la lista de nuevos nombres de columna
    new_col_names.append(name_no_spaces)



print(df.columns)
  # Comprobar el resultado: la lista de encabezados


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-warning">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

Esta parte esta excelente. Podes dejar solo esta celda o ir construyendo como lo hiciste, pero recorda no importar el df.</div>

In [ ]:
OK

[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [100]:
#import pandas as pd

#df = pd.read_csv('/datasets/music_project_en.csv')

new_col_names = []# Encuentra el número de valores ausentes en la tabla
num_valores_ausentes = df.isna().sum()

print(num_valores_ausentes) # Calcular el número de valores ausentes


user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [124]:
columns_to_replace = ['track', 'artist', 'genre']

for col in columns_to_replace:
	df[col].fillna('unknown', inplace=True)# escribe tu código aquí
display(df)
#print(df)
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'


user_id                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina           unknown   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfield  14:07:09     Friday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
4           dance  Springfield  08:34:34     Monday  
...           ...          ...       ...        ...  
65074         rnb  Springfield  13:32:28  Wednesday  
65075         hip  Shelbyville  10:00:00     Monday  
65076  industrial  Springfield  20:09:26     Friday  
65077        rock  Springfield  21:43:59     Friday  
65078     country  Springfield  21:59:46     Friday  

[65079 rows x 7 columns]

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Esta bien. Pero en este caso remplaza el print() y utiliza el metodo sin nada (llamalo solo) o con el metodo display(). Esto hara que sea mas legible y estetico.</div>

<div class="alert alert-block alert-success">

<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido
</div>

OK

Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [102]:
print(df.isna().sum())# Contar valores ausentes


user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


<div class="alert alert-block alert-success">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy bien utilizado el bucle.</div>

[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [103]:
print(df.duplicated().sum())# Contar duplicados explícitos


3826


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [104]:
df = df.drop_duplicates()# Eliminar duplicados explícitos


Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [105]:
print(df.duplicated().sum())# Comprobar de nuevo si hay duplicados



0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [125]:
#print(df['genre'].unique()) # Inspeccionar los nombres de géneros únicos
print(sorted(df['genre'].unique()))  # el nuevo DataFrame sin duplicados

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

En este caso, ordenar estos valores hara que sea mucho mas claro encontrar errores y correcciones. Siempre es mejor para la legibilidad.

Contando la cantidad de elementos que tenemos (antes y despues) tambien es parametro para verificar si el error esta corregido.</div>

<div class="alert alert-block alert-success">

<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido. Muy bien.-
</div>

OK

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [107]:
def replace_wrong_genres(df, column, wrong_genres, correct_genre): # pasar una lista de valores incorrectos y un string con el valor correcto en la entrada de la función
    for wrong_genre in wrong_genres: # un bucle sobre nombres mal escritos
        df[column] = df[column].replace(wrong_genre, correct_genre) # llamar a replace() para cada nombre incorrecto
    return df # devolver el DataFrame modificado


Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [108]:
duplicates = ['hip', 'hop', 'hip-hop'] # una lista de nombres mal escritos
name = 'hiphop' # el nombre correcto


Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [126]:
df = replace_wrong_genres(df, 'genre', duplicates, name) # llamar a la función, replace() se llamará dos veces
#print(df['genre'].unique()) 
print(sorted(df['genre'].unique()))  # el nuevo DataFrame sin duplicados 

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

En este caso, ordenar estos valores hara que sea mucho mas claro encontrar errores y correcciones. Siempre es mejor para la legibilidad.

Contando la cantidad de elementos que tenemos (antes y despues) tambien es parametro para verificar si el error esta corregido.</div>

<div class="alert alert-block alert-success">

<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido
</div>

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`

Al analizar duplicados, seguí los siguientes pasos:
Contar duplicados explícitos: Utilicé el método duplicated() para encontrar y contar duplicados explícitos en el DataFrame.
Eliminar duplicados explícitos: Utilicé el método drop_duplicates() para eliminar los duplicados explícitos y conservar solo la primera ocurrencia de cada entrada duplicada.
Comprobar de nuevo si hay duplicados: Después de eliminar duplicados explícitos, volví a verificar si aún quedaban duplicados en el DataFrame.
Inspeccionar los nombres de géneros únicos: Mostré la lista de nombres de géneros únicos antes y después de corregir los valores incorrectos.
Pasar una lista de valores incorrectos y un string con el valor correcto en la entrada de la función: Identifiqué los valores incorrectos en la lista de géneros y los reemplacé con el valor correcto "hip-hop".
Un bucle sobre nombres mal escritos: Iteré sobre los nombres mal escritos en la lista y llamé a replace() para cada uno, corrigiendo así los errores.
Llamar a replace() para cada nombre incorrecto: Utilicé el método replace() para corregir cada nombre incorrecto en la columna "genre" del DataFrame.
Devolver el DataFrame modificado: El DataFrame fue modificado con las correcciones y sin duplicados explícitos.
Es importante destacar que durante este proceso, se mantuvo la transparencia en las correcciones realizadas y se proporcionó una explicación clara en el código y los comentarios para facilitar la comprensión del análisis y las modificaciones realizadas.


<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Me gusto mucho como lo hiciste en items. En este caso deberas comentar como abordaste cada cosa y los cambios realizados, pero en el texto deberas explicar un poco mas.
    
Recorda que el cliente quiza no entienda el codigo pero si nuestro analisis.


</div>

<div class="alert alert-block alert-success">

<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido. Te quedo muy buena las observaciones.
</div>

OK LISTO

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [110]:
df.groupby(by='city')['user_id'].count()

city
Shelbyville    18512
Springfield    42741
Name: user_id, dtype: int64

`Comenta tus observaciones aquí`

Observaciones sobre la actividad de usuarios y usuarias en cada ciudad:

Ciudades y Número de Usuarios/Usuarias:

Shelbyville tiene un total de 18,512 usuarios/usuarias.
Springfield tiene un total de 42,741 usuarios/usuarias.
Se observa una diferencia significativa en el número de usuarios/usuarias entre las dos ciudades, siendo Springfield considerablemente más grande en términos de usuarios/usuarias.

Canciones Reproducidas por Ciudad:

No se ha especificado la columna que contiene el número de canciones reproducidas, asumiré que es la columna 'user_id'.
La distribución de canciones reproducidas por ciudad es la siguiente:
Shelbyville: 18,512 canciones
Springfield: 42,741 canciones
La diferencia en la cantidad de canciones reproducidas sigue la tendencia del número de usuarios/usuarias, siendo Springfield el líder en ambos aspectos.


<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Recorda hacer tus observaciones explicandole al cliente la distribucion de los datos y a que se puede deber.</div>

<div class="alert alert-block alert-success">

<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido
</div>

OK

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [111]:
df.groupby(by='day')['user_id'].count()# Calcular las canciones reproducidas en cada uno de los tres días


day
Friday       21840
Monday       21354
Wednesday    18059
Name: user_id, dtype: int64

`Comenta tus observaciones aquí`

Observaciones sobre la actividad de usuarios y usuarias por día de la semana:

Canciones Reproducidas por Día:

Viernes: 21,840 canciones
Lunes: 21,354 canciones
Miércoles: 18,059 canciones
Se observa que el viernes es el día con mayor actividad de reproducción, seguido por el lunes y miércoles. Esto podría indicar patrones de comportamiento en la preferencia de escucha durante la semana, siendo el viernes un día destacado en términos de reproducción de canciones.

En resumen, estas observaciones proporcionan una visión general de la actividad de usuarios y usuarias en diferentes contextos, tanto por ciudad como por día de la semana. La cantidad de usuarios, canciones reproducidas y patrones diarios pueden ser indicadores útiles para comprender el comportamiento del público en estas dos ciudades.


<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Recorda hacer tus observaciones explicandole al cliente la distribucion de los datos y a que se puede deber.</div>

<div class="alert alert-block alert-success">

<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido. Perfecto.
</div>

In [ ]:
OK

Estos resultados proporcionan información sobre la distribución de la actividad de escucha a lo largo de la semana. Puedes comparar estos valores para evaluar si hay diferencias significativas en la cantidad de canciones reproducidas en cada uno de estos días. Esto podría contribuir a respaldar o refutar la hipótesis sobre las diferencias en el comportamiento de escucha entre los usuarios y usuarias de las ciudades de Springfield y Shelbyville.

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [127]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day, city):
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    filtered_day = df[df["day"]== day]
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    filtered_city = filtered_day[filtered_day["city"]== city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    results = filtered_city["user_id"].count() #YA CAMBIE 
    # Devolve el número de valores de la columna 'user_id'
    return results


    

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Jorge, esta perfecto como pensaste la funcion.

Hay un cambio para hacer que modifica tus resultados, `filtered_city` fue tu ultima variable modificada con fitlrado. Por lo que esa deberas utilizar para contar el user_id y no la de filtered_day.</div>

<div class="alert alert-block alert-success">

<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido.
</div>

OK

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [128]:
number_tracks('Monday','Springfield')# El número de canciones reproducidas en Springfield el lunes


16715

In [129]:
number_tracks('Monday','Shelbyville')# El número de canciones reproducidas en Shelbyville el lunes


5982

In [130]:
number_tracks('Wednesday','Springfield')# El número de canciones reproducidas en Springfield el miércoles


11755

In [131]:
number_tracks('Wednesday','Shelbyville')# El número de canciones reproducidas en Shelbyville el miércoles


7478

In [132]:
number_tracks('Friday','Springfield')# El número de canciones reproducidas en Springfield el viernes


16890

In [133]:
number_tracks('Friday','Shelbyville')# El número de canciones reproducidas en Shelbyville el viernes


6259

**Conclusiones**

`Comenta si la hipótesis es correcta o se debe rechazar. Explica tu razonamiento.`

Las observaciones realizadas hasta ahora indican que existe una diferencia en la cantidad de canciones reproducidas por los usuarios y usuarias de Shelbyville y Springfield, tanto en general como en días específicos de la semana.

Diferencia General:

Shelbyville tiene 18,512 canciones reproducidas.
Springfield tiene 42,741 canciones reproducidas.
Esto sugiere que hay una diferencia significativa en la cantidad total de canciones reproducidas entre las dos ciudades.

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

`Resume aquí tus conclusiones sobre la hipótesis.`

El análisis realizado proporciona información relevante sobre la actividad de escucha de los usuarios y usuarias en las ciudades de Springfield y Shelbyville durante los días de la semana seleccionados (viernes, lunes y miércoles). Aquí hay algunos comentarios adicionales:

Distribución por Día:

El viernes muestra la mayor cantidad de canciones reproducidas, con 21,840 reproducciones.
El lunes tiene una actividad cercana, con 21,354 reproducciones.
El miércoles tiene la menor cantidad de reproducciones, con 18,059.
Patrones de Escucha:

La información proporciona una visión general de los patrones de escucha a lo largo de la semana.
La actividad más intensa parece estar al comienzo y al final de la semana, siendo el viernes el día con mayor actividad.
Comparación entre Ciudades:

Para evaluar la hipótesis sobre las diferencias en el comportamiento de escucha entre las ciudades, sería necesario comparar los resultados obtenidos para Springfield y Shelbyville.
Próximos Pasos:

Se podrían realizar análisis adicionales, como la comparación de la actividad entre las dos ciudades en los días específicos mencionados.
También sería útil examinar otras variables relevantes, como los géneros de música más reproducidos en cada ciudad.
En general, estos cálculos son un punto de partida para explorar las diferencias en la actividad de escucha y podrían conducir a análisis más detallados según los objetivos específicos de la investigación.


<div class="alert alert-block alert-success">

<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Excelente.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Jorge, muy buen trabajo el que hiciste.

Abarcaste todos los puntos necesarios y a nivel codigo esta bastante prolijo.

Como consejo, no dejes comentado la importacion al df y la carga de librerias ya que no aportan nada a la celda (no constituye una buena practica).

Como te dije el trabajo esta practicamente hecho, restan algunos detalles. Recorda que la conclusion puede cambiar con los nuevos datos.

Saludos.</div>

LISTO GRACIAS

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Jorge, excelentes todas las correcciones.

Tal es asi que te quedo un trabajo perfecto. 

A destacar como fuiste modificando los analisis y te quedaron muy muy buenas, no escatimes en ellos ya que se te da muy bien y agregara valor.

Estuve atento a tu entrega debido a lo que costo el otro trabajo, el trabajo pasa a estar **aprobado**.

Exitos en lo que sigue, saludos y buen fin de semana.</div>

[Volver a Contenidos](#back)